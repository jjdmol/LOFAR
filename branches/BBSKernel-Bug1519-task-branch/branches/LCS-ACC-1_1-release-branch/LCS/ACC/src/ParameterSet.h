//#  ParameterSet.h: ParameterCollectin filled with runtime values.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  Abstract:
//#
//#	 Defines a class the should contain fully filled ParamaterCollections
//#  to be used in runtime to feed the applications with information.
//#
//#  Note: This source is read best with tabstop 4.
//#
//#  $Id$

#ifndef LOFAR_ACC_PARAMETERSET_H
#define LOFAR_ACC_PARAMETERSET_H

// \file ParameterSet.h
// ParameterCollection filled with runtime values.

//# Never #include <config.h> or #include <lofar_config.h> in a header file!
//# Includes
#include <ACC/ParameterCollection.h>

namespace LOFAR {
  namespace ACC {
// \addtogroup ACC
// @{


//# Description of the class.
// The ParameterSet class is a ParameterCollection that is used during runtime
// to feed an application with its runtime values.<br>
// The restrictions of this collections are:<br>
// 1) It must contain a \c versionnr key with a valid versionnumber.
// 2) All values must be filled in.<br>
// This can be checked with the \c check method.
//
// Since ParameterSet is derived from ParameterCollection all manipulation
// methods of this base class are available. The only extra functionality
// ParameterSet offers is the \c check method the checks if the ParameterSet
// is conform the restrictions.
class ParameterSet : public ParameterCollection
{
public:
	// Default constructable;
	ParameterSet();
	~ParameterSet();

	// Define a conversion function from base class to this class
	ParameterSet(const ParameterCollection& that);

	// Allow reading a file for backwards compatibility; OBSOLETE!
	explicit ParameterSet(const string&	theFilename);

	// Copying is allowed.
	ParameterSet(const ParameterSet& that);
	ParameterSet& 	operator=(const ParameterSet& that);

	// Check if the contents of the Parameterset meets the restrictions. The
	// \c errorReport string contains the violations that were found.
	bool check(string&	errorReport) const;

	friend std::ostream& operator<<(std::ostream& os, const ParameterCollection &thePS);

private:
};

// @} addgroup
} // namespace ACC
} // namespace LOFAR

#endif
