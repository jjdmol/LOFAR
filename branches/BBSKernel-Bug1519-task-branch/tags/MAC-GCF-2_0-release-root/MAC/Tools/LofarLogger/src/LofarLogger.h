//#  LofarLogger.h: logger wrapper
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_LOGGER_H
#define LOFAR_LOGGER_H

#include <Common/lofar_string.h>
#include <Common/lofar_map.h>
#include <log4cplus/logger.h>
#include "LofarLoggerNames.h"

// forward declaration

// define convenience macros
#define LOFAR_LOG_TRACE_SCOPE(logger,logEvent) \
    LofarScopeTraceLogger _scopeLogger(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_TRACE(logger,logEvent) \
    LofarLogger::getInstance().trace(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_DEBUG(logger,logEvent) \
    LofarLogger::getInstance().debug(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_INFO(logger,logEvent)  \
    LofarLogger::getInstance().info(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_WARN(logger,logEvent)  \
    LofarLogger::getInstance().warn(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_ERROR(logger,logEvent) \
    LofarLogger::getInstance().error(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);
#define LOFAR_LOG_FATAL(logger,logEvent) \
    LofarLogger::getInstance().fatal(logger,LofarLogger::formatString logEvent ,__FILE__,__LINE__);

/**
 *
 * The LofarLogger class wraps the log4cplus::Logger class
 *
 * Usage:
 *
 *    #define EXAMPLE_LOGGER_NAME "ASTRON.LOFAR.MAC.ExampleLoggerName"
 *
 *    LOFAR_LOG_TRACE(EXAMPLE_LOGGER_NAME,("example %d level message %d","trace",15))
 *    LOFAR_LOG_FATAL(EXAMPLE_LOGGER_NAME,("example fatal level message"))
 *
 * NOTE: the second argument of the macro HAS to be surrounded with brackets.
 */
class LofarLogger
{
    public:

        /**
        * destructor
        */
        virtual ~LofarLogger();

        static LofarLogger getInstance();

        /**
        * logging methods4
        */
        void trace(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void debug(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void info(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void warn(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void error(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);
        void fatal(const string& logger,const string& logEvent,const char* file=NULL,int line=-1);

        log4cplus::Logger& getLogger(const string& logger);
        static string formatString(char* format,...);

    protected:
        /**4
        * constructor. Protected because public construction is not allowed
        */
        LofarLogger();
        LofarLogger(const LofarLogger&);

    private:

        /**
        * Don't allow copying of the LofarLogger object.
        */
        LofarLogger& operator=(const LofarLogger&);

        typedef std::map<const std::string,log4cplus::Logger> TLoggerMap;

        /**
        * keep the loggers in a map for re-use
        */
        TLoggerMap         _loggerMap;
        static LofarLogger _sLofarLogger;
};

/**
 *
 * The LofarScopeTraceLogger class logs trace messages in its constructor and destructor
 *
 */
class LofarScopeTraceLogger
{
    public:
       explicit LofarScopeTraceLogger(const std::string& logger, const std::string& logEvent,const char* file=NULL,int line=-1) :
                    _traceLogger(LofarLogger::getInstance().getLogger(logger),logEvent,file,line)
                {
                };
                ~LofarScopeTraceLogger()
                {
                };

    protected:
                /**
                * constructor. Protected because public construction is not allowed
                */
                LofarScopeTraceLogger();
                LofarScopeTraceLogger(const LofarScopeTraceLogger&);
                LofarScopeTraceLogger& operator=(const LofarScopeTraceLogger&);

    private:
                log4cplus::TraceLogger _traceLogger;
};

#endif
