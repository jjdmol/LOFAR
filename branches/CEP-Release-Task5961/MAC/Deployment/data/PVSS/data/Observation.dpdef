#Observation.dpdef
claim				:Claim

claimPeriod			int
preparePeriod		int
startTime			string
stopTime			string
bandFilter			string
nyquistzone			int
antennaArray		string
receiverList		string
sampleClock			int
runState			string
measurementSet		string
processType		string
processSubtype		string
strategy		string

stationList			string
inputNodeList		string
BGLNodeList			string
storageNodeList		string

Beams.angle1		doubleArr
Beams.angle2		doubleArr
Beams.directionType	stringArr
Beams.subbandList	stringArr

TiedArrayBeams.beamIndex		intArr
TiedArrayBeams.angle1			doubleArr
TiedArrayBeams.angle2			doubleArr
TiedArrayBeams.directionType 	stringArr
TiedArrayBeams.dispersion		doubleArr
TiedArrayBeams.coherent	       	boolArr
