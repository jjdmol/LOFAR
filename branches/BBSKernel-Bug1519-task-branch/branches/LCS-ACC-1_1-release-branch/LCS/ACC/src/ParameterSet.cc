//#  ParameterSet.cc: Implements a map of Key-Value pairs.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#include <ACC/ParameterSet.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_fstream.h>

using namespace std;

namespace LOFAR {
  namespace ACC {

//#-------------------------- creation and destroy ---------------------------
//#
//# Default constructor
//#
ParameterSet::ParameterSet()
{}

//#
//# ParameterSet(filename)
//#
ParameterSet::ParameterSet(const string&	filename) :
	ParameterCollection(filename)
{}

//#
//# ParameterSet (ParameterCollection)
//#
ParameterSet::ParameterSet(const ParameterCollection& that) :
	ParameterCollection(that)
{}

//#
//# Copying is allowed.
//#
ParameterSet::ParameterSet(const ParameterSet& that)
  : ParameterCollection (that)
{}

//#
//# operator= copying
//#
ParameterSet& ParameterSet::operator=(const ParameterSet& that)
{
	if (this != &that) {
		ParameterCollection::operator= (that);
	}
	return (*this);
}

//#
//#	Destructor
//#
ParameterSet::~ParameterSet()
{}

//#
//# operator<<
//#
std::ostream&	operator<< (std::ostream& os, const ParameterSet &thePS)
{
	return os << static_cast<ParameterCollection> (thePS);
}

//#
//# check()
//#
//# 1) Versionnumber must be present
//# 2) All keys must be filled.
//#
bool	ParameterSet::check(string&	errorReport) const
{
	bool	isOk = true;
	errorReport  = "";

	if (getVersionNr() == "") {
		errorReport += "No versionnumber found\n";
		isOk = false;
	}

	const_iterator		iter = begin();
	while (iter != end()) {
		if (iter->second == "") {
			errorReport += "Key " + iter->first + " has no value\n";
			isOk = false;
		}
		iter++;
	}

	return (isOk);
}

} // namespace ACC
} // namespace LOFAR
