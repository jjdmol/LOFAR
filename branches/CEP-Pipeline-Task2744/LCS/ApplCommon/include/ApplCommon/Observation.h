//# Observation.h: class/struct that holds the Observation information
//#
//# Copyright (C) 2006
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

#ifndef LOFAR_APPLCOMMON_OBSERVATION_H
#define LOFAR_APPLCOMMON_OBSERVATION_H

// \file
// class/struct that holds the Observation information

//# Never #include <config.h> or #include <lofar_config.h> in a header file!

//# Includes
#include <Common/ParameterSet.h>
#include <Common/lofar_bitset.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_string.h>
#include <Common/LofarConstants.h>

namespace LOFAR {

// \addtogroup ApplCommon
// @{

//# Forward Declarations
//class forward;


// The Observation class is a public struct that can be used as base-class
// for holding Observation related information.
// It can be instantiated with a parset containing Observation information.
class Observation
{
public:
	Observation();
	~Observation();
	explicit	Observation (ParameterSet*		aParSet, bool	hasDualHBA = false);

	// global function for converting filtername to nyquist zone
	static uint nyquistzoneFromFilter(const string&	filterName);

	// get name of a beam (idx starts at 0)
	string getBeamName(uint	beamIdx) const;
	string getAnaBeamName() const;

	// check if the given Observation conflicts with this one
	bool	conflicts(const	Observation&	other) const;

	// Returns a bitset containing the RCU's requested by the observation.
	bitset<MAX_RCUS> getRCUbitset(int nrLBAs, int nrHBAs, const string& anAntennaSet);
	// TEMP HACK
	string getAntennaArrayName(bool hasSplitters) const;

	// Support for dynamic dataslot allocation
	vector<int>	getBeamAllocation(const string&	stationName = "") const;
	vector<int>	getBeamlets (uint beamIdx, const string&	stationName = "") const;

	// for operator <<
	ostream& print (ostream&	os) const;

	// data types
	typedef bitset<MAX_RCUS> 	  RCUset_t;

	class Pointing {
	public:
		Pointing() {};
		~Pointing() {};
		double			angle1;
		double			angle2;
		string			directionType;
		time_t			startTime;
		int				duration;
	};
		
	class Beam {
	public:
		Beam() {};
		~Beam() {};
		Beam& operator=(const Beam& that) {
			if (this != &that) {
				this->name 		 = that.name;
				this->antennaSet = that.antennaSet;
				this->pointings  = that.pointings;
				this->momID 	 = that.momID;
				this->subbands 	 = that.subbands;
			}
			return (*this);
		}

		string				name;
		string				antennaSet;
		vector<Pointing>	pointings;

		int					momID;
		vector<int>			subbands;
	};

	class AnaBeam {
	public:
		AnaBeam() {};
		~AnaBeam() {};
		AnaBeam& operator=(const AnaBeam& that) {
			if (this != &that) {
				this->name 		 = that.name;
				this->antennaSet = that.antennaSet;
				this->pointings  = that.pointings;
				this->rank 		 = that.rank;
			}
			return (*this);
		};

		string				name;
		string				antennaSet;
		vector<Pointing>	pointings;

		int					rank;
	};

    class StreamToStorage {
    public:
        string dataProduct;
        unsigned streamNr;
        string filename;

        unsigned sourcePset;
        string destStorageNode;
        string destDirectory;
          
    };

	//# Datamembers
	string			name;
	int				obsID;
	time_t			startTime;
	time_t			stopTime;
	uint			nyquistZone;
	vector<string>	stations;
	int				nrSlotsInFrame;
	int				sampleClock;		// 160 | 200
	string			filter;				// LBA_30_80, LBA_10_90, HBA_110_190, etc.
	string			MSNameMask;
	string			realPVSSdatapoint;

	// old way of specifying antennas
	string			antennaArray;

	// new way of selecting antennas
	string			antennaSet;			// like LBA_INNER, LBA_OUTER, etc.
	bool			useLongBaselines;
	bool			splitterOn;			// On or Off
	bool			dualMode;			// HBA_DUAL selected

	// beams
	vector<Beam>	beams;
	vector<AnaBeam>	anaBeams;

    vector<StreamToStorage> streamsToStorage; 

	// couple of values of the virtual instrument as compacted strings
	string			receiverList;
	string			stationList;
	string			BGLNodeList; 	 
	string			storageNodeList;

private:
	RCUset_t		RCUset;				// set with participating receivers, use getRCUbitset to get this value.

	// many(!) vectors for dataslot allocation
	bool			itsHasDataslots;	// old or new type of parset
	bool			itsStnHasDualHBA;	// 
	ParameterSet	itsDataslotParset;	// subset of parset with dataslotinfo for getxxxAllocation()
	vector<int>		itsSlotTemplate;	// clean template with the slots that may be used.
	vector<uint>	itsBeamSlotList;	// beamnumber vector
	vector<int>		beamlet2beams;		// OLD:to which beam each beamlet belongs
};

//#
//# operator <<
//#
inline ostream& operator<< (ostream&	os, const Observation&	anObs)
{
	return (anObs.print(os));
}

// @}
} // namespace LOFAR

#endif
