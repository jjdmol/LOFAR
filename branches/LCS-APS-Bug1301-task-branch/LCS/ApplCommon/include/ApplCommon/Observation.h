//#  Observation.h: class/struct that holds the Observation information
//#
//#  Copyright (C) 2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#ifndef LOFAR_APPLCOMMON_OBSERVATION_H
#define LOFAR_APPLCOMMON_OBSERVATION_H

// \file
// class/struct that holds the Observation information

//# Never #include <config.h> or #include <lofar_config.h> in a header file!

//# Includes
#include <Common/ParameterSet.h>
#include <Common/lofar_bitset.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_string.h>
#include <Common/LofarConstants.h>

namespace LOFAR {

// \addtogroup ApplCommon
// @{

//# Forward Declarations
//class forward;


// The Observation class is a public struct that can be used as base-class
// for holding Observation related information.
// It can be instantiated with a parset containing Observation information.
class Observation
{
public:
	Observation();
	~Observation();
	explicit	Observation (ACC::APS::ParameterSet*		aParSet);

	// global function for converting filtername to nyquist zone
	static uint32 nyquistzoneFromFilter(const string&	filterName);

	// get name of a beam (idx starts at 0)
	string getBeamName(uint32	beamIdx) const;

        // OLAP: get the subbandList
	vector<uint32> getSubbandList() const;
        // OLAP: get the beamList
	vector<uint32> getBeamList() const;
        // OLAP: get the rspBoardList
	vector<uint32> getRspBoardList() const;
        // OLAP: get the rspSlotList
	vector<uint32> getRspSlotList() const;
 
	// for operator <<
	ostream& print (ostream&	os) const;

	// data types
	typedef bitset<MAX_RCUS> 	  RCUset_t;

	class Beam {
	public:
		Beam() {};
		~Beam() {};

		// NOTE: since not other sw in the signal chain supports switching the beam to another direction
		//		 we only support 1 direction per beam for now.
		double			angle1;
		double			angle2;
		string			directionType;
//		string			angleTimes;
		vector<int32>	subbands;
		vector<int32>	beamlets;
	};

	//# Datamembers
	string			name;
	int32			obsID;
	time_t			startTime;
	time_t			stopTime;
	uint32			nyquistZone;
	vector<string>	stations;
	int32			sampleClock;
	string			filter;
	string			MSNameMask;
	string			realPVSSdatapoint;

	// old way of specifying antennas
	string			antennaArray;
	RCUset_t		RCUset;				// set with participating receivers

	// new way of selecting antennas
	string			antennaSet;
	bool			useLongBaselines;

	vector<Beam>	beams;
	vector<int32>	beamlet2beams;		// to which beam each beamlet belongs
	vector<int32>	beamlet2subbands;	// which subband each beamlet uses.
	uint32	        nrRSPboards;	        // the number of RSPboads.
	vector<uint32>	subbandList;	        // OLAP: subband list.
	vector<uint32>	beamList;	        // OLAP: beam list.
	vector<uint32>	rspBoardList;           // OLAP: RSP board list.
	vector<uint32>	rspSlotList;            // OLAP: RSP Slot list
	bool            uStation;               // use uStation mode?

	// couple of values of the virtual instrument as compacted strings
	string			receiverList;
	string			stationList;
	string                  BGLNodeList; 	 
	string                  storageNodeList;
};

//#
//# operator <<
//#
inline ostream& operator<< (ostream&	os, const Observation&	anObs)
{
	return (anObs.print(os));
}

// @}
} // namespace LOFAR

#endif
