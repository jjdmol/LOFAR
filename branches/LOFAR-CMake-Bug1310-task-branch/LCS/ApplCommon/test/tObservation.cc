//# tObservation.cc
//#
//# Copyright (C) 2002-2004
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <Common/LofarLogger.h>
#include <Common/ParameterSet.h>
#include <ApplCommon/Observation.h>

using namespace LOFAR;

int main (int, char* argv[]) 
{
	INIT_LOGGER(argv[0]);
	try {
		ParameterSet parSet1("tObservation.in_parset1");
		Observation  obs1(&parSet1);
		cout << obs1 << endl;

		// add an extra beam
		parSet1.replace("ObsSW.Observation.nrBeams", "2");
		parSet1.add("ObsSW.Observation.Beam[1].angle1", 		"0.23456789");
		parSet1.add("ObsSW.Observation.Beam[1].angle2", 		"0.123456789");
		parSet1.add("ObsSW.Observation.Beam[1].directionType",  "AZEL");
		parSet1.add("ObsSW.Observation.Beam[1].subbandList", 	"[4,3,102]");
		parSet1.add("ObsSW.Observation.Beam[1].beamletList", 	"[15,16,18]");
		Observation  obs2(&parSet1);
		cout << obs2 << endl;

		cout << ">>>" << endl;
		// test conflicts in clock
		ParameterSet conflictPS1("tObservation.in_conflict1");
		Observation  conflictObs1(&conflictPS1);
		ASSERTSTR(obs2.conflicts(conflictObs1), "File 1 should have had a clock conflict");
	
		// test conflicts in receivers
		ParameterSet conflictPS2("tObservation.in_conflict2");
		Observation  conflictObs2(&conflictPS2);
		ASSERTSTR(obs2.conflicts(conflictObs2), "File 2 should have had a receiver conflict");
	
		// test conflicts in beamlets
		ParameterSet conflictPS3("tObservation.in_conflict3");
		Observation  conflictObs3(&conflictPS3);
		ASSERTSTR(obs2.conflicts(conflictObs3), "File 3 should have had a beamlet conflict");
	
		// test conflicts in nrSlotsPerFrame
		ParameterSet conflictPS4("tObservation.in_conflict4");
		Observation  conflictObs4(&conflictPS4);
		ASSERTSTR(obs2.conflicts(conflictObs4), "File 4 should have had a nrSlotInFrame conflict");
	
		// everything conflict except the time
		ParameterSet conflictPS5("tObservation.in_conflict5");
		Observation  conflictObs5(&conflictPS5);
		ASSERTSTR(!obs2.conflicts(conflictObs5), "File 5 should NOT have had a conflict");
		cout << "<<<" << endl;
		cout << "No conflict found in file 5 which is oke." << endl;

		// test RCUbitset based on receiverList
		bitset<MAX_RCUS>	expectedRCUs;
		expectedRCUs.reset();
		for (int r = 0; r < 12; r++) {
			expectedRCUs.set(r);
		}
		cout << "getRCUbitset(48,48,12,false) = " << obs1.getRCUbitset(48,48,12,false) << endl;

		// basic test on RCU bitsets
		parSet1.replace("ObsSW.Observation.antennaSet", "LBA_OUTER");
		Observation		obs3(&parSet1);
		cout << obs3.antennaSet << endl;
		cout << "getRCUbitset(96,48,12,true)  = " << obs3.getRCUbitset(96,48,12,true) << endl;	// Core
		cout << "getRCUbitset(96,48,12,false) = " << obs3.getRCUbitset(96,48,12,false) << endl;	// Remote
		cout << "getRCUbitset(96,48,24,false) = " << obs3.getRCUbitset(96,48,24,false) << endl;	// Europe
		cout << "getRCUbitset(96,96,24,false) = " << obs3.getRCUbitset(96,96,24,false) << endl;	// Europe
		
		// basic test on RCU bitsets
		obs3.antennaSet = "HBA_BOTH";
		cout << obs3.antennaSet << endl;
		cout << "getRCUbitset(96,48,12,true)  = " << obs3.getRCUbitset(96,48,12,true) << endl;	// Core
		cout << "getRCUbitset(96,48,12,false) = " << obs3.getRCUbitset(96,48,12,false) << endl;	// Remote
		cout << "getRCUbitset(96,48,24,false) = " << obs3.getRCUbitset(96,48,24,false) << endl;	// Europe
		cout << "getRCUbitset(96,96,24,false) = " << obs3.getRCUbitset(96,96,24,false) << endl;	// Europe
		
		// tricky test on RCU bitsets
		obs3.antennaSet = "HBA_ONE";
		cout << obs3.antennaSet << endl;
		cout << "getRCUbitset(96,48,12,true)  = " << obs3.getRCUbitset(96,48,12,true) << endl;	// Core
		cout << "getRCUbitset(96,48,12,false) = " << obs3.getRCUbitset(96,48,12,false) << endl;	// Remote
		cout << "getRCUbitset(96,48,24,false) = " << obs3.getRCUbitset(96,48,24,false) << endl;	// Europe
		cout << "getRCUbitset(96,96,24,false) = " << obs3.getRCUbitset(96,96,24,false) << endl;	// Europe
		
		cout << "HBA_ONE(false) = " << obs3.getAntennaArrayName(false) << endl;
		cout << "HBA_ONE(true)  = " << obs3.getAntennaArrayName(true) << endl;
		obs3.antennaSet = "HBA_TWO";
		cout << "HBA_TWO(false) = " << obs3.getAntennaArrayName(false) << endl;
		cout << "HBA_TWO(true)  = " << obs3.getAntennaArrayName(true) << endl;
		obs3.antennaSet = "HBA_BOTH";
		cout << "HBA_BOTH(false) = " << obs3.getAntennaArrayName(false) << endl;
		cout << "HBA_BOTH(true)  = " << obs3.getAntennaArrayName(true) << endl;

		obs3.antennaSet = "LBA_INNER";
		cout << "LBA_INNER(false) = " << obs3.getAntennaArrayName(false) << endl;
		cout << "LBA_INNER(true)  = " << obs3.getAntennaArrayName(true) << endl;
		obs3.antennaSet = "LBA_OUTER";
		cout << "LBA_OUTER(false) = " << obs3.getAntennaArrayName(false) << endl;
		cout << "LBA_OUTER(true)  = " << obs3.getAntennaArrayName(true) << endl;
		obs3.antennaSet = "LBA_X";
		cout << "LBA_X(false) = " << obs3.getAntennaArrayName(false) << endl;
		cout << "LBA_X(true)  = " << obs3.getAntennaArrayName(true) << endl;
	}
	catch (Exception& e) {
		cout << "Exception: " << e.what() << endl;
		return 1;
	}
	return 0;
}
