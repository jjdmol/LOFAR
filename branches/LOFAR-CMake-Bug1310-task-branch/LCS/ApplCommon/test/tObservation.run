#! /bin/sh

tObservation tObservation.in_parset1
