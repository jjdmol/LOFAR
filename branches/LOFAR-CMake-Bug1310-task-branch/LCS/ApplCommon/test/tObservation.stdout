
Observation  : 
ObsID        : 0
starttime    : 2008-Dec-16 14:30:00
stoptime     : 2008-Dec-16 14:34:00
stations     : [CS016,CS017,CS018]
antennaArray : HBA
receiver set : 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001
sampleClock  : 160
filter       : HB_210_240
nyquistZone  : 3

Meas.set     : /data/L${YEAR}_${MSNUMBER}/SB${SUBBAND}.MS

Receivers    : [0]
Stations     : [CS016..CS018]
BLG nodes    : [bgl001..bgl003]
Storage nodes: [list001,list002]

nrBeams      : 1
Beam[0].pointing   : 6.000000, 0.500000, 
Beam[0].subbandList: [2,3,4,5]
Beam[0].beamletList: [0,1,2,3]
beamlet2beams   : [0,0,0,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]
beamlet2subbands: [2,3,4,5,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1]

uStation        : False
nrRSPboards     : 1
subbandList     : [2,3,4,5]
beamList        : [0,0,0,0]
rspBoardList    : [0,0,0,0]
rspSlotList     : [0,1,2,3]

